python/ray/data/_internal/planner/planner.py

```diff
@@ -1,28 +1,98 @@
-from ray.data._internal.planner.plan_all_to_all_op import plan_all_to_all_op
-from ray.data._internal.planner.plan_from_op import plan_from_op
-from ray.data._internal.planner.plan_input_data_op import plan_input_data_op
-from ray.data._internal.planner.plan_limit_op import plan_limit_op
-from ray.data._internal.planner.plan_read_op import plan_read_op
-from ray.data._internal.planner.plan_udf_map_op import plan_udf_map_op
-from ray.data._internal.planner.plan_write_op import plan_write_op
+from ray.util.annotations import DeveloperAPI
+
+LogicalOperatorType = TypeVar("LogicalOperatorType", bound=LogicalOperator)
+
+PlanLogicalOpFn = Callable[
+    [LogicalOperatorType, List[PhysicalOperator]], PhysicalOperator
+]
+
+# A list of registered plan functions for logical operators.
+PLAN_LOGICAL_OP_FNS: List[Tuple[Type[LogicalOperator], PlanLogicalOpFn]] = []
+
+
+@DeveloperAPI
+def register_plan_logical_op_fn(
+    logical_op_type: Type[LogicalOperator],
+    plan_fn: PlanLogicalOpFn,
+):
+    """Register a plan function for a logical operator type."""
+    PLAN_LOGICAL_OP_FNS.append((logical_op_type, plan_fn))
+
+
+def register_plan_logical_op_fns():
+    from ray.data._internal.execution.operators.input_data_buffer import InputDataBuffer
+    from ray.data._internal.execution.operators.limit_operator import LimitOperator
+    from ray.data._internal.execution.operators.union_operator import UnionOperator
+    from ray.data._internal.execution.operators.zip_operator import ZipOperator
+    from ray.data._internal.logical.operators.all_to_all_operator import (
+        AbstractAllToAll,
+    )
+    from ray.data._internal.logical.operators.from_operators import AbstractFrom
+    from ray.data._internal.logical.operators.input_data_operator import InputData
+    from ray.data._internal.logical.operators.map_operator import AbstractUDFMap
+    from ray.data._internal.logical.operators.n_ary_operator import Union, Zip
+    from ray.data._internal.logical.operators.one_to_one_operator import Limit
+    from ray.data._internal.logical.operators.read_operator import Read
+    from ray.data._internal.logical.operators.write_operator import Write
+    from ray.data._internal.planner.plan_all_to_all_op import plan_all_to_all_op
+    from ray.data._internal.planner.plan_read_op import plan_read_op
+    from ray.data._internal.planner.plan_udf_map_op import plan_udf_map_op
+    from ray.data._internal.planner.plan_write_op import plan_write_op
+
+    register_plan_logical_op_fn(Read, plan_read_op)
+
+    def plan_input_data_op(
+        logical_op: InputData, physical_children: List[PhysicalOperator]
+    ) -> PhysicalOperator:
+        """Get the corresponding DAG of physical operators for InputData."""
+        assert len(physical_children) == 0
+
+        return InputDataBuffer(
+            input_data=logical_op.input_data,
+            input_data_factory=logical_op.input_data_factory,
+        )
```

Comment:
Nit: IMO this'll be easier to read if we consistently define plan functions in separate modules. I think mixing the two approaches (function definitions in separate modules and in this function) hurts readability.

Not heavily opinionated on this, so am okay to keep as is.

Reply:
some plan functions are just one line. I feel it's simpler and cleaner to just put them here. For those complex plan functions, it's better to put them in separate files.